uber/automated_emails.py

```diff
@@ -1,75 +1,290 @@
+        1) whether we have sent this exact email out yet or not (previously_sent_emails)
+        2) whether the email category has been approved (approved_idents)
+        3) whether the model instance passed in is the same type as what we want to process
+        4) do any date-based filters exist on this email category? (i.e. send 7 days before magfest)
+        5) do any other filters exist on this email category? (i.e. only if attendee.staffing == true)
+
+        Example #1 of a model instance to check:
+          self.ident: "You {attendee.name} have registered for our event!"
+          model_inst:  class Attendee: id #4532, name: "John smith"
+
+        Example #2 of a model instance to check:
+          self.ident: "Your group {group.name} owes money"
+          model_inst:  class Group: id #1251, name: "The Fighting Mongooses"
+
+        :param model_inst: The model we've been requested to use (i.e. Attendee, Group, etc)
+
+        :return: True if we should send this email to this model instance, False if not.
+        """
+
+        if not isinstance(model_inst, self.model):
+            return False
+
+        if not model_inst.email:
+            return False
+
+        if self._already_sent(model_inst):
+            return False
+
+        if not self.filters_run(model_inst):
+            return False
+
+        if not self.approved:
+            return False
+
+        return True
 
-    def render(self, x):
-        model = getattr(x, 'email_model_name', x.__class__.__name__.lower())
-        return render('emails/' + self.template, dict({model: x}, **self.extra_data))
+    @property
+    def approved(self):
+        """
+        Check if this email category has been approved by the admins to send automated emails.
 
-    def send(self, x, raise_errors=True):
+        :return: True if we are approved to send this email, or don't need approval. False otherwise
+        """
+
+        approved_to_send = not self.needs_approval or self.ident in c.EMAIL_APPROVED_IDENTS
+
+        if not approved_to_send:
+            # log statistics about how many emails would have been sent if we had approval.
+            # if running as part of a daemon, this will record the data.
+            SendAllAutomatedEmailsJob.log_unset_because_unapproved(self)
+
+        return approved_to_send
+
+    def render(self, model_instance):
+        model = getattr(model_instance, 'email_model_name', model_instance.__class__.__name__.lower())
+        return render('emails/' + self.template, dict({model: model_instance}, **self.extra_data))
+
+    def really_send(self, model_instance):
+        """
+        Actually send an email to a particular model instance (i.e. a particular attendee).
+
+        Doesn't perform any kind of checks at all if we should be sending this, just immediately sends the email
+        no matter what.
+
+        NOTE: use send_if_possible() instead of calling this method unless you 100% know what you're doing.
```

Comment:
Would it be better to make this is a private function, since we don't really want people using it outside this class?